Basic UI

DmT021 opened the pull request.

Commit:
.gitignore added

Commit:
.gitignore updated

Commit:
basic UI implemented

mrdekk commented on NotesV1/AppDelegate.swift:
```diff
@@ -0,0 +1,119 @@
+
+    var window: UIWindow?
+
+    static var sharedInstance: AppDelegate {
```
sharedInstance - это из соглашений по наименованию Obj-C и swift до 3 версии
Начиная со Swift 3 лучше писать просто shared
https://github.com/apple/swift-evolution/blob/master/proposals/0005-objective-c-name-translation.md

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTable.swift:
```diff
@@ -0,0 +1,201 @@
+    func fillCells() {
+        if dataSource != nil {
+            //let section = 0 // TODO get support for multiple sections
+            let rowsCount = dataSourceTyped!.numberOfRows(self)
```
force unwrapping optionals (dataSourceTyped!) опасная штука
вообще весь этот if лучше написать так

``` swift
if let ds = dataSourceTyped {
   let rowsCount = ds.numberOfRows(self)
  ...
```

Так принято в swift

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTable.swift:
```diff
@@ -0,0 +1,201 @@
+    required init?(coder aDecoder: NSCoder) {
+        selectedIndex = -1
+        super.init(coder: aDecoder)
+        //dataSource = aDecoder.decodeObject(forKey: "dataSource") as? MyUIHorizontalTableDataSource
```
Код который не используется и не будет использоваться в будущем принято удалять (он в общем останется в истории изменений и если вдруг понадобится - его есть где взять и вернуть)
Ежели хочется его по какой-то причине оставить - лучше добавить комментарий, почему этот код стоит оставить

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTable.swift:
```diff
@@ -0,0 +1,201 @@
+            //let section = 0 // TODO get support for multiple sections
+            let rowsCount = dataSourceTyped!.numberOfRows(self)
+            for r in 0..<rowsCount {
+                let cell = createCell()
```
Опять к force unwrapping (вызов optional через !) - это приведет к падению приложения если вдруг объект все-таки nil поэтому лучше сделать так

``` swift
if let cell = createCell {
  cell.index = r
  cellViews.append(cell)
}
```

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTable.swift:
```diff
@@ -0,0 +1,201 @@
+
+            self.stack.subviews.forEach({ v in v.removeFromSuperview() })
+            for r in 0..<rowsCount {
+                let cell = dataSource?.cell(self, forRowAt: r)
```
``` swift
if let cell = dataSource?.cell(self, forRowAt: r) {
  self.stack.addArrangedSubview(cell)
```

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTable.swift:
```diff
@@ -0,0 +1,201 @@
+
+    // cell factory
+    func createCell() -> MyUIHorizontalTableCell? {
+        if cellClass != nil {
```
``` swift
guard let cellClass = cellClass else { return nil }
let cell = cellClass.init()
...
```

Тут стоит помнить про golden path https://github.com/raywenderlich/swift-style-guide/issues/63

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTable.swift:
```diff
@@ -0,0 +1,201 @@
+        if index >= 0 && index < cellViews.count {
+            return cellViews[index]
+        }
+        // TODO is exception would be better here?
```
no

иначе надо сигнатуру сделать так

``` swift
func getCellForIndex(index: Int) -> MyUIHorizontalTableCell throws
```

а вообще согласно Grand Renaming в Swift 3 лучше сделать вообще так

``` swift
func table(_ table: MyUIHorizontalTable, cellFor index: Int) -> MyUIHorizontalTableCell?
```

или даже

``` swift
func cell(for index: Int) -> MyUIHorizontalTableCell?
```

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTable.swift:
```diff
@@ -0,0 +1,201 @@
+        return nil
+    }
+
+    func cellTapped(_ cell: MyUIHorizontalTableCell) {
```
``` swift
func didTap(_ cell: MyUIHorizontalTableCell) ...
```

?

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTableCell.swift:
```diff
@@ -0,0 +1,64 @@
+
+@objc
+protocol MyUIHorizontalTableCellDelegate {
+    @objc optional func cellTapped(_ cell: MyUIHorizontalTableCell)
```
``` swift
... func didTap(_ cell: MyUIHorizontalTableCell)
```

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTableCell.swift:
```diff
@@ -0,0 +1,64 @@
+@IBDesignable
+class MyUIHorizontalTableCell: UIView, UIGestureRecognizerDelegate {
+    private var defaultColor: UIColor?
+    private var tapRecognizer: UITapGestureRecognizer!
```
Где-то у Apple в доках были описаны причины, почему не стоит размещать UIGestureRecognizer'ы непосредственно в элементах UITableView и UICollectionView но я на вскидку найти не могу.
Поэтому пусть тут будет так - не вижу каких-то явных проблем, но об этом стоит помнить.

mrdekk commented on NotesV1/MyUIHorizontalTable/MyUIHorizontalTableCell.swift:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @objc private func tapRecognizerAction() {
+        if delegate != nil {
```
``` swift
if let delegate = delegate {
  delegate.didTap(self)
}
```

mrdekk commented on NotesV1/MyUISegmentedColorSelector/MyUISegmentedColorSelector.swift:
```diff
@@ -0,0 +1,88 @@
+}
+
+@objc
+class ColorsListBasedOnArray: NSObject, MyUIHorizontalTableDataSource {
```
ColorListDataSource ? (не ColorsListBasedOnArray)

mrdekk commented on NotesV1/MyUISegmentedColorSelector/MyUISegmentedColorSelector.swift:
```diff
@@ -0,0 +1,88 @@
+}
+
+@objc
+class ColorsList: NSObject, MyUIHorizontalTableDataSource {
```
RandomColorListDataSource ?

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+class NoteDetailViewController: UIViewController, UITextViewDelegate,
+    MyUISegmentedColorSelectorDelegate {
+
+    let serviceLocator = AppDelegate.sharedInstance.serviceLocator!
```
Очень не хочется использовать singleton'ы, пусть он и один и это AppDelegate
В этом приложении пусть будет так, но в будущем стоит обратить внимание на IoC контейнеры

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        var visibleRect = self.view.frame
+        visibleRect.size.height -= keyboardFrame.height
+        let textEditingFrame = getTextEditingFrame()
+        if textEditingFrame != nil {
```
``` swift
if let frame = textEditingFrame {
  if !visibleRect.contains(frame.origin) {
    scrollView.scrollRectToVisible(frame, animated: true)
  }
}
```

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    private func getTextEditingFrame() -> CGRect? {
+        if textEditingControl?.selectedTextRange != nil {
```
``` swift
guard let control = textEditingControl, let range = control.selectedTextRange else { return nil }
let caretRect = control.caretRect(for: range.start)
let frame = self.scrollView.convert(caretRect, from: control as? UIView)
return frame
```

Проблемы в имеющемся коде:
1. force unwrapping optionals
2. именование переменных одной буквой - так делать не стоит
3. golden path

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+            title: "Delete",
+            message: "Are you sure?",
+            preferredStyle: .actionSheet)
+        let deleteAction = UIAlertAction(title: "Delete", style: .destructive, handler: { act in
```
closure захватывает self сильной ссылкой, с учетом того, что сам alert можен быть захвачен self'ом ссильной ссылкой - можем получить retain cycle. completion'ы и handler'ы можно делать после скобок

лучше сделать так

``` swift
let deleteAction = UIAlertAction(title: "Delete", style: .destructive) { [weak self] act in
  guard let sself = self else { return } // sself - strong self
  sself.deleteConfirmed()
}
```

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    private func deleteConfirmed() {
+        if noteId != nil {
```
``` swift
if let noteId = noteId {
  serviceLocator.dataService.removeNote(noteId: noteId)
}
```

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    @IBAction
+    private func saveTapped(_ sender: Any) {
+        // copy data from cotrols into dataObject
+        dataItem?.colorId = colorSelector.selectedIndex
```
может стоит сделать так

``` swift
if let item = dataItem {
  item.colorId = ...
  ...
}
```

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        dataItem?.desc = descriptionField.text
+        dataItem?.title = titleField.text
+
+        if noteId == nil {
```
``` swift
if let noteId = noteId {
  serviceLocator.dataService.updateNote(noteId: noteId, note: item) // как получили item см выше
} else {
  serviceLocator.dataService.addNote(note: item)
}
```

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+
+        let n = Note()
+        n.colorId = 0
+        let ts = Date().description(with: Locale.current)
```
Для преобразования объектов типа Date в строку лучше использовать DateFormatter

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        if let note = serviceLocator.dataService.getNote(noteId: noteId) {
+            dataItem = note
+        } else {
+            // TODO handle error. Should we show an error message to the user?
```
yes

mrdekk commented on NotesV1/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    private func configureView() {
+        if titleField == nil ||
```
guard titleField != nil, descriptionField != nil, colorSelector != nil else { return }

mrdekk commented on NotesV1/NoteListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+class NoteListViewController: UIViewController {
+
+    @IBOutlet weak var tableView: UITableView!
+    private var notesListGeneration: Int = 0
```
про generation и то как обойтись без него - обсуждали 10 ноября вечерком

mrdekk commented on NotesV1/NoteListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        super.viewDidLoad()
+        // Do any additional setup after loading the view, typically from a nib.
+
+        tableView.dataSource = notesListDataSource
```
dataSource и delegate лучше назначать через Storyboard

mrdekk commented on NotesV1/NoteListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        tableView.delegate = notesListDataSource
+    }
+
+    override func didReceiveMemoryWarning() {
```
Если ничего не делается в таких функциях - лучше их не писать совсем

mrdekk commented on NotesV1/NoteListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+
+    private let serviceLocator = AppDelegate.sharedInstance.serviceLocator!
+    private let dataService: IDataService
+    private var gen: Int = -1
```
Лучше убрать

mrdekk commented on NotesV1/NoteListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    private let serviceLocator = AppDelegate.sharedInstance.serviceLocator!
+    private let dataService: IDataService
+    private var gen: Int = -1
+    private var prefetchedNotes: InvalidatableComputedValue<[Note]>!
```
Как обойтись без InvalidatableComputedValue обсуждали 10 ноября вечером
Лучше обойтись простым массивом

mrdekk commented on NotesV1/NoteListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+    }
+}
+
+class NoteListViewCell: UITableViewCell {
```
Такие классы лучше выносить в отдельный файл
Не надо их все собирать здесь

mrdekk commented on NotesV1/NoteListViewController.swift:
```diff
@@ -0,0 +1,139 @@
+        gen = dataService.getNotesGeneration()
+    }
+
+    func checkNeedsUpdate() -> Bool {
```
не надо

mrdekk commented on NotesV1/Services/DataService.swift:
```diff
@@ -0,0 +1,85 @@
+
+import Foundation
+
+class Note: NSObject {
```
Лучше выносить модельные классы в отдельные файлы

mrdekk commented on NotesV1/Services/DataService.swift:
```diff
@@ -0,0 +1,85 @@
+    var colorId: Int = 0
+}
+
+protocol IDataService : class, NSObjectProtocol {
```
И это тоже
Добавлять префикс I к протоколам в Swift и ObjC не принято

mrdekk commented on NotesV1/Utils/InvalidatableComputedValue.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+class InvalidatableComputedValue<TValue> {
```
Такой подход лучше не использовать

mrdekk reviewed (changes requested):
Посмотрел
По комментариям отписался

Commit:
AppDelegate.sharedInstance was renamed to just shared
unused code was removed
force unwrapping was replaced with if let and guard let
MyUIHorizontalTable.fillCells now produces error when dataSource isn't configured
naming convention for: MyUIHorizontalTableCellDelegate.didTap (was cellTapped), MyUIHorizontalTable.cell (was getCellForIndex), ColorsListDataSource (was ColorsListBasedOnArray)
The handler of deleteAction in NoteDetailViewController.deleteTapped doesn't capture strong self anymore
String representation of Date is produced by DateFormatter now (not Date().description)
NoteDetailViewController.openAsEdit now returns success state. This is a temporal behavior. The logic will be moved into parent VC
NoteListViewController.didReceiveMemoryWarning overriding was removed
if openAsEdit fails alert with a error message will be shown

Commit:
swiftlint configuration file added

Commit:
model classes is now in Model directory
IDataService was renamed to DataServiceProtocol and moved into file DataServiceProtocol.swift

Commit:
fix missing IDataService in NoteListDataSource
NoteListViewCell was moved into NoteListViewCell.swift

Commit:
Data update logic was moved from NoteDetailViewController to NoteListViewController
tableView reloads only selected item
DataService is not support generations of data
DataService.addNote returns id of added note
NoteListDataSource doesn't contain cache of items anymore
NoteDetailViewController has dataObjectUpdated to let parent VC know when the note updated

Commit:
InvalidatableComputedValue was removed

Commit:
ColorsListDataSource and RandomColorsListDataSource was moved into ColorsListDataSource.swift
NoteListDataSource was moved into NoteListDataSource.swift

Commit:
tableView.delegate and tableView.dataSource is assigned by the storyboard now

DmT021 committed:
fix for crash on iPad when delete note button tapped

DmT021 committed:
Merge branch 'hotfix/ipad_crash'

mrdekk commented:
OK
